[Feature] 프로필 UI 수정

dudtjs1021ej opened the pull request:
### 📕 Issue

x


### 📙 작업 내역

- [x] 내 프로필의 게시물 정방형으로 수정 (FlowLayout -> CompositionalLayout으로 구현)
- [x] 내 프로필 메뉴 수정
- [x]  내 프로필, 게시물 미니 프로필 구현
- [x] 내 게시물 수정, 삭제 추가
- [x] UIColor rgb값에서 공통 컬러셋으로 수정


### 📘 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트


### 📋 체크리스트

- [x] Merge 하는 브랜치가 올바른가?
- [x] PR과 관련없는 변경사항이 없는가?
- [x] 내 코드에 대한 자기 검토가 되었는가?
- [ ] 변경사항이 효과적이거나 동작이 작동한다는 것을 보증하는 테스트를 추가하였는가?
- [ ] 새로운 테스트와 기존의 테스트가 변경사항에 대해 만족하는가?

### 📱 테스트 방법


- 내 게시물 수정 클릭 시 핏프티 등록으로 이동합니다.
- 내 게시물 삭제 클릭 시 프로필탭으로 이동합니다.
- 유저의 게시물의 미니 프로필 클릭 시 유저 프로필로 이동합니다.

### 📝 리뷰 노트

- 내 프로필에서 게시글 셀로 스크롤하면 보이는 미니 프로필뷰
  -> 네비게이션바를 사용하려 했으나 height를 늘릴 수 없어 UIView로 구현하였습니다.
- 게시물 삭제, 게시물 수정시 화면 전환 방식이 괜찮은지 봐주세요!
- 
### 🏞️ 스크린샷
<img src = "https://user-images.githubusercontent.com/77915491/214854561-f2a227c8-6a50-4af1-a1ae-677dc477c81b.png" width="30%"> <img src = "https://user-images.githubusercontent.com/77915491/214854428-5e63c894-0f76-4f96-9e24-1ff4e2e866d5.png" width="30%">

dudtjs1021ej committed:
💄 [style] 내 프로필 게시물 사진 정방형으로 수정

dudtjs1021ej committed:
💄 [style] 내 프로필 텍스트버튼 제거

dudtjs1021ej committed:
✨ [feat] MiniProfileView 구현

dudtjs1021ej committed:
✨ [feat] MyPostBottomSheetView 구현

dudtjs1021ej committed:
💄 [style] MiniProfileView 높이 변경

- NavigationBar에서 UIView로 변경

dudtjs1021ej committed:
✨ [feat] 게시물 미니 프로필 구현

dudtjs1021ej committed:
✨ [feat] User 게시물 미니 프로필 구현

dudtjs1021ej committed:
✨ [feat] 게시물 수정, 삭제 화면 전환 구현

- 게시물 수정 시 핏프티 등록으로 이동
- 게시물 삭제 시 내 프로필로 이동

dudtjs1021ej committed:
💄 [style] 미니 프로필 구분선 추가

dudtjs1021ej committed:
♻️ [refactor] 내 프로필 header height 프로퍼티로 관리

dudtjs1021ej committed:
💄 [style] UIColor 공통 컬러셋으로 수정

leeari95 commented on Fitfty/Projects/Coordinator/Sources/Profile/ProfileCoordinator.swift:
```diff
@@ -36,12 +36,59 @@ private extension ProfileCoordinator {
+        navigationController.present(coordinator.navigationController, animated: true)
+    }
+    
+    func dismiss() {
+        navigationController.dismiss(animated: false)
+    }
```
이부분 개선이 필요할 것 같아요.
메모리 누수때문에 꼭 필요해요. 현재 코드에서는 바텀시트를 닫아도 메모리에 계속 살아있을거에요.
참고하시라고 코드 예시를 알려드릴게요!

```swift
// AddressViewController.swift 32번 라인
    override public func removeFromParent() {
        super.removeFromParent()
        coordinator?.dismiss() // removeFromParent가 호출 시 코디네이터의 dismiss 메소드를 호출하도록 재정의
    }
```
```swift
// AddressCoordinator.swift 45번 라인...
    func dismiss() {
        navigationController.viewControllers.removeAll()
        bottomSheetDelegate?.dismissBottomSheet() // delegate를 통해 바텀시트를 dismiss
        finishDelegate?.coordinatorDidFinish(childCoordinator: self) 
    }
```
```swift
// MainCoordinator.swift 38번라인
    func makeAddressViewController() -> UIViewController {
        let coordinator = AddressCoordinator()
        coordinator.parentCoordinator = self
        childCoordinators.append(coordinator)
        coordinator.start()
        coordinator.finishDelegate = self
        coordinator.parentCoordinator = self
        let bottomSheetViewController = BottomSheetViewController(
            style: .large,
            contentViewController: coordinator.navigationController
        )
        coordinator.bottomSheetDelegate = bottomSheetViewController // 바텀시트를 닫기 위한 delegate
        return bottomSheetViewController
    }
```

dudtjs1021ej replied:
몰랐던 부분이네요..!! 예시코드까지 감사해여👍 참고해서 수정해보겠습니다 ㅎㅎ

leeari95 commented on Fitfty/Projects/Profile/Sources/MyProfileCoordinatorInterface.swift:
```diff
@@ -6,10 +6,13 @@
 //
 
-import Foundation
+import UIKit
```
[P5] import UIKit이 필요한 이유가 있나용?

dudtjs1021ej replied:
앗 테스트 해본다고 파라미터로 UIViewController 넣어봤다가 같이 들어갔네여😅 제거하겠습니다!

leeari95 reviewed (approved):
썬~ 수정사항 확인했습니다! 고생많으셨어요~~
코드를 살펴보았는데, 특이사항은 딱히 없었네요! 
다만 구조적인 부분에 있어서 몇가지 저의 의견들을 좀 적어볼게요!!

## 저의 의견

### - MyProfile, UserProfile, MyPost, UserPost

`내 프로필, 유저프로필`과 `내 포스트, 유저포스트` 화면이 
몇부분을 제외하고는 모두 `공통되는 뷰`인 것 같아요. 
지금은 시간관계상 어렵겠지만, 나중에라도 프로필, 포스트를 합쳐서 
각 뷰컨트롤러에 모드에 따라 (`user`, `my`) 특정 뷰를 보여주거나, 가릴 수 있도록 
리팩토링을 해주면 코드가 좀더 간결해질 것 같아요.
모드에 따라 화면을 달리 보여주도록 하면
하나의 뷰 컨트롤러를 여러 곳에서 재사용할 수 있게 될 것 같아요!

### - MyPostBottomSheetView

게시물 수정, 삭제와 같은 뷰는 확장성을 생각해서
테이블뷰로 구현을 해두면, 더 좋지 않을까라는 생각도 드네요!
지금 상태로는 뭔가 메뉴가 추가되었을 때 버튼을 계속 구현해줘야하는 반면에
테이블뷰 같은 걸로 구현해두었을 때에는, 셀을 추가해줄 수 있도록 모델만 추가해주면 되니까...
이 부분도 코드로 봤을 때 더 간결해지면서 수정도 간편해지지 않을까 하는 생각이 들었어요.

#

제가 남겨드린 의견들은 지금 당장 반영하기 어렵다고 생각이 드신다면
당장 반영하시지 않아도 괜찮아요. 다만 썬이 참고해주시면 좋겠다고 생각이 들어서 말씀드려보아요!
이번에도 구현해주시느라 고생 많으셨습니다!! 😊🙏🏻

dudtjs1021ej reviewed (commented):
@leeari95 
아리 리뷰 감사합니다!!~~

@MyProfile, UserProfile, MyPost, UserPost
저도 이 부분 구현하면서 비효율적인 것 같다고 생각했는데 딱 말해주셨어여..👍
아리가 말해주신 방법으로 리팩토링 진행하면 더 간결해질 것 같네요..!! 
다만 시간적 여유가 부족할 것 같아 우선순위를 좀 미루겠습니다😂

@MyPostBottomSheetView
테이블뷰로 구현하는게 확장성 면에서 좋을 것 같네요!! 요건 금방 구현할 수 있을 것 같아서 이번에 반영해보겠습니다ㅎㅎ

+디자인 수정사항이 추가되어서 이 부분도 같이 구현해 커밋하겠습니다💪

dudtjs1021ej committed:
🔥 [remove] 불필요한 import 코드 제거

dudtjs1021ej committed:
🐛 [fix] 바텀시트 메모리 누수 개선

dudtjs1021ej committed:
🚚 [style] StyleTag, WeatherTag Common으로 이동

dudtjs1021ej committed:
💄 [style] PostInfoView UI 수정

dudtjs1021ej committed:
💄 [style] BookmarkView 추가

dudtjs1021ej committed:
💄 [style] PostInfoView spacing 수정, eyeGray05 이미지 추가

dudtjs1021ej committed:
✨ [feat] ProfileType 추가

dudtjs1021ej committed:
♻️ [refactor] ProfileViewController, PostViewController를 Profile 모드로 관리

- MyPostViewController, UserPostViewController -> PostViewController로 관리
- MyProfileViewController, UserProfileViewController -> ProfileViewController로 관리
- PostCoordinator 추가

dudtjs1021ej committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/profile-ios2-129

dudtjs1021ej committed:
🚚 [style] Profile 폴더 구조 수정

dudtjs1021ej committed:
✨ [feat] Profile PresentType 추기

dudtjs1021ej committed:
✨ [feat] PresentType에 따라 ProfileViewController navigationBar 처리

dudtjs1021ej committed:
✨ [feat] 미니 프로필 탭 할 시 화면 전환 추가

dudtjs1021ej committed:
💄 [style] 북마크 CommonAsset 이미지로 변경

dudtjs1021ej committed:
✨ [feat] PostViewController ScrollView 구현

dudtjs1021ej committed:
💄 [style] BtnCheckBoxSelected, BtnCheckBoxUnSelected 이미지 추가

dudtjs1021ej committed:
✨ [feat] DetailReportViewController 구현

- ReportView, AllReportView 추가

dudtjs1021ej committed:
✨ [feat] DetailReportCoordinator 구현, DetailReportViewController UI 추가

dudtjs1021ej committed:
✨ [feat] 설정 화면 전환 추가

dudtjs1021ej committed:
🐛 [fix] PostViewController 네비게이션바 타이틀 보이는 현상 해결

dudtjs1021ej committed:
✨ [feat] 신고 사유 체크 기능 추가

dudtjs1021ej committed:
💄 [style] 내비게이션바 컬러 white로 고정

- ProfileViewController 네비게이션바 밑줄 제거

dudtjs1021ej commented on Fitfty/Projects/Common/Sources/AppAppearance.swift:
```diff
@@ -13,5 +13,6 @@ public final class AppAppearance {
         UIBarButtonItem.appearance().tintColor = .label
         UINavigationBar.appearance().largeTitleTextAttributes = [.foregroundColor: UIColor.label]
         UINavigationBar.appearance().titleTextAttributes = [.foregroundColor: UIColor.label]
+        UINavigationBar.appearance().barTintColor = .white
```
스크롤하면 navigationBar 색이 바뀌길래 white로 고정해주었어요!

dudtjs1021ej commented on Fitfty/Projects/MainFeed/Sources/CoordinatorInterfaces/MainCoordinatorInterface.swift:
```diff
@@ -5,16 +5,15 @@
 //
-
-import Foundation
+import Profile
```
여기서 xclode cloud 에러 뜨는데...해결을 못하고 있어요😂 혹시 뭐가 문제인지 아시나여...?

leeari95 replied:
import Foundation 때문인지 알았는데, 제 착각이였네요.
MainFeed는 Profile을 의존하고 있지 않아서 생기는 문제같아요.
저 프로필 타입을 Common 모듈로 이동시키는 건 어떤가요?
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/75905803/215778643-902a436d-7d78-4605-95fd-b83677c060dd.png">

dudtjs1021ej replied:
아하..!! 감사합니다!! 옮겨볼게용

dudtjs1021ej replied:
아직 체크중이긴한데 해결된 것 같네요..!! 의존성 문제였군여 감사해요 아리~!!🥰

dudtjs1021ej commented on Fitfty/Projects/Profile/Sources/PresentType.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+public enum PresentType {
+    case main
+    case tab
+}
```
탭에서 프로필을 들어가면 설정, 메뉴가 보여야하고 
메인의 코디탭에서 들어가면 보이지 않아야해서 PresentType으로 구분하도록 해주었어요..!

dudtjs1021ej commented on Fitfty/Projects/Profile/Sources/ProfileType.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+public enum ProfileType {
+    case myProfile
+    case userProfile
+}
```
유저프로필인지 내프로필인지, 유저게시물인지 내 게시물인지
구분하기 위해 ProfileType을 만들었습니당

dudtjs1021ej commented on Fitfty/Projects/Setting/Sources/Home/ViewController/SettingViewController.swift:
```diff
@@ -59,6 +59,7 @@ private extension SettingViewController {
             action: #selector(didTapBackButton(_:))
         )
         navigationItem.leftBarButtonItem = cancelButton
+        navigationController?.navigationBar.isHidden = false
```
프로핍탭에서 navigationBar를 hidden처리해서 다시 보이게 추가했습니다!

dudtjs1021ej reviewed (commented):
@leeari95 

아리! 전에 주신 피드백 생각해봤는데 ViewModel까지 구현되면 유저와 내 프로필 리팩토링이 더 힘들어질 것 같아 이번에 구현해보았습니다..!
생각보다 시간이 좀 걸려서 MyPostBottomSheetView 리팩토링은 다음에 진행해보도록 할게여🥹

- __UserPostVC, MyPostVC -> PostVC
  UserProfileVC, MyProfileVC -> ProfileVC__
    - `profileType`, `presentType`으로 한 뷰컨에서 구분할 수 있도록 구현했습니당
 
    
<br>
<br>

<img src = "https://user-images.githubusercontent.com/77915491/215772476-c6aac5c6-5bb5-4790-99a4-126dfd9bd819.png" width="30%">

-  __PostVC__

    -  글자수 제한이 늘어나서 게시글에 스크롤뷰를 추가했습니다.
    -  피그마 변경사항을 반영해 UI를 수정했습니다!
    
<br>    
<br>

<img src="https://user-images.githubusercontent.com/77915491/215773599-8ff24015-1690-4ed2-b468-f070a0fdb188.png" width="30%">

- __DetailReportVC 구현__
    - 유저 신고 클릭 시 나오는 DetailReportVC를 구현했습니다😄
    
    
+추가로 클라우드에서 Profile 모듈을 못찾는데...뭐가 문제일까요..🥲

leeari95 committed:
♻️ [refactor] import Foundation 추가

leeari95 commented on Fitfty/Projects/Profile/Sources/Views/Post/PostView.swift:
```diff
@@ -20,20 +21,36 @@ final class PostView: UIView {
+        button.setImage(CommonAsset.Images.bookmarkFill.image, for: .normal)
+        button.setPreferredSymbolConfiguration(.init(scale: .large), forImageIn: .normal)
+        return button
+    }()
```
[P4] 북마크 버튼을 탭했을 때, 북마크 설정/해제 효과도 추가해주면 좋을 것 같아요!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/169c4ce8a8a5489d1ca4494dd0a4ca7f1f770e51

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/PostViewController.swift:
```diff
@@ -0,0 +1,154 @@
+                
+            stackView.topAnchor.constraint(equalTo: scrollView.topAnchor),
+            stackView.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor),
+            stackView.widthAnchor.constraint(equalTo: scrollView.widthAnchor)
```
[P3] 썬~ 스크롤뷰 레이아웃 디버깅 돌리니까 경고문구가 나오네요! 🤔

<img width="2090" alt="image" src="https://user-images.githubusercontent.com/75905803/215780838-58db5f86-1741-479a-87bd-1b985840d55e.png">

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/909e7661de0b0c212208ecb29e0c76860d8401e8

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/PostViewController.swift:
```diff
@@ -0,0 +1,154 @@
+        }
+    }
+    
+    private func showNavigationBar() {
+        if presentType == .tab {
+            navigationController?.navigationBar.isHidden = true
+        }
+    }
```
[P3] 메소드 네이밍이 적절하지 않은 것 같아요. 보여주다가 아니라 숨긴다는 의미는 어떨까용?
그리고 if문 보다는 guard문이 더 적절해보여요!

dudtjs1021ej replied:
엇 그러네요..!! 왜 show로 했을까요..ㅎㅎ 수정하겠습니다! 
제 생각은 presentType이 tab일때를 체크하는거라 
조건이 아닐때를 체크하는 guard문보다 if문이 나을 것 같은데 어떻게 생각하시나여..?!

leeari95 replied:
사실 if문이든 guard문이든 별 상관은 없는 것 같은데...ㅋㅋㅋ 
이것도 개발자들 사이에서 말이 많은 주제인 것 같아요.

일단 제 의견을 공유드리자면....☺️
제가 guard문이 더 적절하다고 생각했던 이유는, 
presentType이 tab이 아닐 때 함수를 실행하지 않는 방향이 더 가독성이 좋다고 생각했기 때문이에요.

저는 if가 있으면 else도 왠지 있어야할 것 같은 느낌이 들어요. 여러 조건에 따라 분기처리를 할 때 주로 사용하니까요.

guard는 조건이 맞지 않을 때 항상 return이나 break을 통해 exit 시키는 방향이잖아요? 
코드 depth 측면에서도... 더 좋다는 생각이 들었구요.
guard문이 가독성이 향상된다는 근거 내용은 [Swift Language Guide](https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html#ID525)에도 기재되어 있어요!

썬 의견도 궁금하네용 🤔

dudtjs1021ej replied:
오옹 if문이 있으면 else문이 있어야할 것 같기도 하네요...!
다만 지금은 presentType이 두가지라 상관없을 것 같은데 타입이 늘어나는 걸 고려해서
 제 생각엔 guard문보단 밑에 아리가 제시해준 switch문으로 처리해볼까 하는데 어떠신가요?!

leeari95 replied:
넵! switch문도 좋은 것 같습니다 ☺️👍

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/107d47d469e39ae3a86eed866f4f1a9fa0595d9c

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/PostViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    private func setNavigationBar() {
+        navigationController?.navigationBar.topItem?.title = ""
+        navigationController?.navigationBar.prefersLargeTitles = false
+        if presentType == .tab {
+            navigationController?.navigationBar.isHidden = false
+        }
+        if profileType == .myProfile {
+            navigationController?.navigationItem.setCustomRightBarButton(
+                self,
+                action: #selector(didTapRightBarButton),
+                image: CommonAsset.Images.btnMoreVertical.image,
+                size: 24
+            )
+        }
```
[P4] switch 문으로 개선할 수 있을 것 같아요!

dudtjs1021ej replied:
presentType이 main일때와 profileType이 userProfile일때는 체크할 필요가 없어서 if문으로 작성해봤습니다..!!
switch문으로 작성하면 불필요한 코드가 늘어나는 느낌인데 개선해야 할까요..?😯

leeari95 replied:
만약 추후에 userProfile에도 별도로 분기처리가 필요할 일이 생긴다면, 그때도 if로 분기처리를 진행하는게 맞을까요?
switch문이라면 기존 `default` 케이스를 제거하고 `case .userProfile`을 추가해주어서 간단하게 수정할 수 있을 것 같아요.

enum처럼 케이스별로 분기처리를 해야하는 상황이라면 switch가 가독성이 더 좋아보여서 의견드려보았습니다 😊

dudtjs1021ej replied:
지금 당장은 처리할 일이 없어 생각못했는데 추후를 생각하면 switch문으로 처리하는게 낫겠네요~
의견 감사합니다! 반영해볼게요😄

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/107d47d469e39ae3a86eed866f4f1a9fa0595d9c

leeari95 commented on Fitfty/Projects/Coordinator/Sources/Profile/PostCoordinator.swift:
```diff
@@ -0,0 +1,120 @@
+    
+    init(navigationConrtoller: BaseNavigationController = BaseNavigationController()) {
+        self.navigationController = navigationConrtoller
+        navigationConrtoller.setCustomBackButton()
```
[P2] 네비게이션 컨트롤러의 바버튼 설정을 코디네이터에서 해주고 있는데, 
코디네이터 역할에 맞는 로직일까요? 한번 더 고민해보셨으면 좋겠어요! 😊
그리고, 현재 이 설정으로 backButton을 누를 경우, 
코디네이터가 메모리 해제가 되지 않아서 인스턴스가 계속 메모리에 누적되는 구조에요.
개선이 필요해보여요!
(MainCoordinator.swift 114번 라인도 같이 봐주세요!)

>한가지 저도 썬 코드 리뷰하면서 발견하게 되었는데, 백 제스처로 pop해도 코디네이터 제거가 안되요.
그래서 ViewController가 deinit 되는 시점에도 
coordinator가 메모리에서 해제될 수 있도록 추가 로직이 필요할 것 같아요.
제가 작업했던 곳은 제가 개선해볼게요. 참고해주세요!

```swift
    deinit {
        coordinator.finished()
    }
```

leeari95 replied:
저의 경우는 커스텀한 backButton을 추가해주고, action을 따로 구현해주었어요.

```swift
// WeatherViewController.swift 66번 라인 ...
func setUpNavigationBar() {
    let cancelButton = UIBarButtonItem(
        image: CommonAsset.Images.btnArrowleft.image,
        style: .plain,
        target: self,
        action: #selector(didTapBackButton(_:))
    )
    navigationItem.leftBarButtonItem = cancelButton
}

@objc func didTapBackButton(_ sender: UITapGestureRecognizer) {
    coordinator.finished()
}
```

dudtjs1021ej replied:
메모리 누수 체크 못하고있었네요ㅠㅠ 개선해보겠습니다!!💪

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/7eeeb0ee5858a321093af77e67b41c8a19945880

dudtjs1021ej committed:
🚚 [style] PresentType, PresentType Common으로 파일 이동

leeari95 commented on Fitfty/Projects/Profile/Sources/Views/Report/ReportView.swift:
```diff
@@ -0,0 +1,59 @@
+        } else {
+            checkBoxButton.isSelected = true
+        }
+    }
```
[P2] 그냥 뷰 자체를 누를 때는 select가 되는데, 체크 버튼을 누를 경우에는 체크가 제대로 동작을 안하는 것 같아요!
개선이 필요해보입니당!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/bb26ac31bb457cc8a20872a3e5ec826e518f03af

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/MyPostBottomSheetViewController.swift:
```diff
@@ -0,0 +1,57 @@
+            myPostBottomSheetView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20),
+            myPostBottomSheetView.topAnchor.constraint(equalTo: view.topAnchor, constant: 44),
+            myPostBottomSheetView.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -44)
+        ])
```
[P2] 보니까 이 바텀시트 높이를 정적으로 196 사이즈를 주고 있는 것 같네요.
그럼 이 bottomAnchor 제약조건을 제거하고, 
MyPostBottomSheetView.swift 내부 setConstraintsLayout()에 spacing을 24로 할당하는 방법은 어떨까요?
그럼 바텀시트가 지금 설정한거보다 더 자연스럽게 올라올 것 같아요.

>이전에 말씀드렸던 부자연스러운 부분을 개선하기 위해서 
바텀 시트 레이아웃 잡을 때는 bottomAnchor 기준으로 레이아웃을 잡는 것을 피하는 게 좋을 것 같아요.

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/1815e0f5852c10b19fcf51c0e06d66e9b63c7a7a

dudtjs1021ej commented on Fitfty/Projects/Common/Sources/ProfilePresentType.swift:
```diff
@@ -0,0 +1,14 @@
+public enum ProfilePresentType {
+    case main
+    case tab
+}
```
Common으로 옮기면서 PresentType->ProfilePresentType으로 수정했는데 네이밍 괜찮나요..?😅

leeari95 replied:
[P4] 네 더 명확하니 좋은 것 같아요. 👍
근데 main, tab 이라는 네이밍 보다 나의 프로필, 다른 사용자의 프로필 이런식의 네이밍이 더 명확할 것 같아요!

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/ReportViewController.swift:
```diff
@@ -11,12 +11,12 @@ import Common
 
 final public class ReportViewController: UIViewController {
     
```
[P2] 이 뷰컨에 아래 코드도 추가 부탁드립니다.
설명드리자면, 바텀시트를 쓸어 내리거나, 어두운 부분을 탭했을 때, 호출되는 메소드에요.
코디네이터를 메모리에서 정상적으로 해제하기 위한 작업입니다!

```swift
override public func removeFromParent() {
    super.removeFromParent()
    coordinator.dismiss()
}
```

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/DetailReportViewController.swift:
```diff
@@ -0,0 +1,66 @@
+import Common
+
+final public class DetailReportViewController: UIViewController {
+    
```
[P2] 이 뷰컨에도 아래 코드 추가 부탁드립니다!!

```swift
override public func removeFromParent() {
    super.removeFromParent()
    coordinator.dismiss()
}
```

leeari95 commented on Fitfty/Projects/Coordinator/Sources/Profile/ProfileCoordinator.swift:
```diff
@@ -44,24 +69,85 @@ private extension ProfileCoordinator {
+        return coordinator
+    }
+    
+    func makeReportViewController() -> UIViewController {
+        let bottomSheetViewController =
+        BottomSheetViewController(
+            style: .small,
+            contentViewController: ReportViewController(coordinator: self)
+        )
+        bottomSheetDelegate = bottomSheetViewController
+        return bottomSheetViewController
+    }
```
[P5] 왜 ReportViewController는 코디네이터가 존재하지 않나요?

dudtjs1021ej replied:
신고뷰는 ProfileCoordinator에서만 쓰일 것 같아 코디네이터로 따로 구현하지 않았던 것 같아요..!
재사용을 고려하면 Coordinator 구현을 하는게 나을까요?

leeari95 replied:
딱 이 친구만 코디네이터가 없는 것 같아 특별한 이유가 있는지 궁금해서 여쭤본 것이였어요. 
나중에 필요로 할 때 구현해도 상관은 없지만, 일관성있게 구현해두면 좋지 않을까? 했어요 ㅎㅎ
이 부분은 썬이 판단하셔도 괜찮을 것 같아요.

leeari95 commented on Fitfty/Projects/Profile/Sources/Views/Report/ReportView.swift:
```diff
@@ -0,0 +1,59 @@
+        button.setImage(CommonAsset.Images.btnCheckBoxSelected.image, for: .selected)
+        button.widthAnchor.constraint(equalToConstant: 20).isActive = true
+        return button
+    }()
```
[P2] 흠 뭔가... 버튼이 찌그러진 느낌이 든다고 생각이 들어서 확인해봤는데
높이 레이아웃도 같이 설정해줘야할 것 같아요!! 
> 아래는 피그마 설정값과 xcode 레이아웃 디버깅 캡쳐화면이에요 ㅎㅎ
<img width="505" alt="image" src="https://user-images.githubusercontent.com/75905803/215790481-c1d40943-ced7-4cca-8aa2-3cc83c0124e3.png">
<img width="259" alt="image" src="https://user-images.githubusercontent.com/75905803/215790564-8a10d1fd-a3cb-48e6-bef3-7a6a9bce4262.png">

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/71c453c55045413fb1e163027caa77acc7b81948

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -0,0 +1,264 @@
+        let section = NSCollectionLayoutSection(group: group)
+        section.contentInsets = .init(top: 27, leading: 12, bottom: 5, trailing: 12)
+        
+        switch presentType {
+        case .main:
+            section.boundarySupplementaryItems = [
+                NSCollectionLayoutBoundarySupplementaryItem(
+                    layoutSize: .init(
+                        widthDimension: .absolute(UIScreen.main.bounds.width),
+                        heightDimension: .estimated(headerHeight)
+                    ),
+                    elementKind: UserProfileHeaderView.className,
+                    alignment: .top
+                )
+            ]
+            
+        case .tab:
+            section.boundarySupplementaryItems = [
+                NSCollectionLayoutBoundarySupplementaryItem(
+                    layoutSize: .init(
+                        widthDimension: .absolute(UIScreen.main.bounds.width),
+                        heightDimension: .estimated(headerHeight)
+                    ),
+                    elementKind: MyProfileHeaderView.className,
+                    alignment: .top
+                )
+            ]
+        }
```
[P3] 이렇게 개선해보면 어떨까요? 훨씬 간결해질 것 같아요 ㅎㅎ
```swift
section.boundarySupplementaryItems = [
    NSCollectionLayoutBoundarySupplementaryItem(
        layoutSize: .init(
            widthDimension: .absolute(UIScreen.main.bounds.width),
            heightDimension: .estimated(headerHeight)
        ),
        elementKind: presentType == .main ? UserProfileHeaderView.className : MyProfileHeaderView.className,
        alignment: .top
    )
]
```

dudtjs1021ej replied:
오옹 확실히 더 짧아지네요..👍
간결성 측면에서 너무 좋긴한데 main인 경우, main이 아닌경우 두가지로만 나뉘어서  presentType이 늘어날때를 대비할 수 없을 것 같아요🥲
코드 간결성이냐 확장성이냐 문제같은데 고민되네요...🤔
아리 의견은 어떠신가여..?!

leeari95 replied:
음... 그러면 저 부분을 함수로 분리해서
파라미터로 presetType을 전달받고, 
해당 타입에 따라 switch를 타서 
각각 유형에 맞는 elementKind를 반환해줄 수도 있을 것 같아요!

아니면 presentType 내부에 헤더뷰의 elementKind를 반환하는 프로퍼티를 만드는 방법도 괜찮을 것 같아요 ㅎㅎ

dudtjs1021ej replied:
> presentType 내부에 헤더뷰의 elementKind를 반환하는 프로퍼티를 만드는 방법

너무 좋네요..!!! 이렇게 구현해보도록 할게요👍

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/674a0ac513b37e3f1dff567880b91bcff15dfaea

이건 PresentType이 Common모듈에 있어서 Profile모듈에 있는 UserProfileHeaderView.className이 접근이 안되더라구요🥲
그래서 ProfileViewController 내부에 elementKind프로퍼티를 만들었어요..!

leeari95 commented on Fitfty/Projects/Profile/Sources/Views/Report/AllReportView.swift:
```diff
@@ -0,0 +1,40 @@
+        ReportView(title: "지적재산권 침해"),
+        ReportView(title: "혐오 / 욕설 / 인신공격"),
+        ReportView(title: "같은 내용 반복 게시"),
+        ReportView(title: "기타 (team.fitfty@gmail.com로 제보)")
```
[P5] 이것도 StackView 보다는 나중에 테이블뷰나 컬렉션뷰로 개선해보면 좋을 것 같습니다 🤗

leeari95 replied:
[P2] 레이아웃 경고가 존재해요!

<img width="538" alt="image" src="https://user-images.githubusercontent.com/75905803/215795600-e423bb6a-391a-43e4-a64e-391df7cdd651.png">

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/8fe2313f4836b302094b885e787cd0e66c3e529f

leeari95 commented on Fitfty/Projects/Profile/Sources/Views/Report/ReportView.swift:
```diff
@@ -0,0 +1,59 @@
+            checkBoxButton.isSelected = false
+        } else {
+            checkBoxButton.isSelected = true
+        }
```
[P4] 이렇게도 할 수 있어요! 😆
```swift
checkBoxButton.isSelected = checkBoxButton.isSelected ? false : true
```

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/f769e4372c886be5b2876b55ce2181c1fca5c2b0

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -0,0 +1,264 @@
+        case .tab:
+            return 283
+        }
+    }
```
[P5] 이 부분은 ProfilePresentType 내부 프로퍼티로 구현해도 되지 않을까요?

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/674a0ac513b37e3f1dff567880b91bcff15dfaea

leeari95 commented on Fitfty/Projects/Profile/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -0,0 +1,264 @@
+
+final public class ProfileViewController: UIViewController {
+    
+    enum Section: CaseIterable {
+        case feed
+    }
```
[P3] 썬! 이 부분은 저도 에단한태 리뷰 받았던 내용인데, 참고해주시면 좋을 것 같아요.
싱크를 맞추면 좋겠다 싶어서 공유 드립니다 🙏🏻
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/12#discussion_r1072090354

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/17/commits/395be85159727f652dc627c23939279442c2f275

leeari95 reviewed (changes requested):
와우 썬... 리팩토링 하시느라 고생 많으셨습니다!!!!
다만 제가 몇가지 문제들을 발견해서 코멘트 달아두었습니다.
수정하시다가 어려움 있으시면 편하게 디엠 주세용 😊🙏🏻

자세한건 코멘트를 달아두었고, 전체적인 내용을 정리하자면 아래와 같습니다.

## 제가 발견한 문제

* Coordinator 부분 메모리 누수 문제
	* backButton 클릭 혹은 백제스처 시 Coordinator 메모리 해제 안됨
	* 바텀시트를 쓸어내리거나, 뒷 부분을 탭했을 때 메모리 해제 안됨
* 레이아웃 경고 문제
	* `AllReportView`의 레이아웃 경고
	* PostViewController 내부 `scrollView` 레이아웃 경고

## 저의 의견

* 뷰를 구성할 때 네이밍이 명확하지 않은 것들이 존재해요.
	* stackView, scrollView 등등... 어떤 스택뷰인지 어떤 스크롤뷰인지 이름만으로 알기 어려워요.
	* 코드를 자세히 확인해야 해당 뷰의 용도를 파악할 수 있어서 정확한 네이밍을 해서 개선해주는 것이 좋을 것 같아요. 😊🙏🏻
* Coordinator 문제 때문에 메모리 누수가 발생하고 있어요.
	* 일단 제가 문제라고 생각되는 부분을 상세히 코멘트를 달아두었는데, 모두 개선하고 누수가 더 발생하지는 않는지 디버깅이 필요해보입니다! 🥹

> 요거는 제가 누수라고 의심되는 인스턴스들을 체크해보았어요. 참고부탁드립니다!
<img width="694" alt="image" src="https://user-images.githubusercontent.com/75905803/215799094-898c6df2-3bc2-4f29-a9f3-f1ef543ae60a.png">

#

제가 꼼꼼히 보겠다고 신경써서 보긴 했는데... 혹시나 더 필요한 부분이 있거나, 궁금한 점이 있다면
언제든지 자유롭게 코멘트 남겨주시거나, 디엠 주시면 감사하겠습니다.
썬 화이팅~~!! 😁💪💪💪

dudtjs1021ej committed:
🐛 [fix] PostViewController scrollView 레이아웃 에러 해결

dudtjs1021ej committed:
🚚 [style] ProfilePresentType mainProfile, tabProfile로 네이밍 변경

dudtjs1021ej committed:
🐛 [fix] 신고 사유 체크버튼 선택 안됨 해결

dudtjs1021ej committed:
🐛 [fix] AllReportView 레이아웃 에러 해결

dudtjs1021ej committed:
💄 [style] checkBoxButton height 20으로 추가

dudtjs1021ej committed:
♻️ [refactor] ReportView isSelected 체크 코드 개선

dudtjs1021ej committed:
♻️ [refactor] ProfilePresentType, 프로필 headerViewElementKind 개선

dudtjs1021ej committed:
♻️ [refactor] PostViewController 개선

- switch문으로 ProfileType, PresentType 체크
- 적절하지 않은 메소드명 hideNavigationBar로 변경

dudtjs1021ej committed:
♻️ [refactor] MyPostBottomSheetView present 자연스럽게 개선

dudtjs1021ej committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/profile-ios2-129

dudtjs1021ej committed:
♻️ [refactor] ProfileCoordinator, PostCoordinator 메모리 누수 개선

dudtjs1021ej committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/profile-ios2-129

dudtjs1021ej committed:
♻️ [refactor] dismissBottomSheet 수정

dudtjs1021ej committed:
🔧 [chore] ProfilePresentType import UIKit 추가

dudtjs1021ej committed:
🐛 [fix] 계속 pop 되는 현상 해결

dudtjs1021ej committed:
🚚 [style] Profile 모듈 Profile과 Post 폴더로 구분

dudtjs1021ej committed:
♻️ [refactor] ProfileViewController 내부 Section을 ViewModel 안에 ProfileSection으로 변경

dudtjs1021ej committed:
✨ [feat] PostView 북마크 탭 구현

dudtjs1021ej committed:
🐛 [fix] dismissBottomSheet 모든 VC 사라지는 현상 해결

dudtjs1021ej committed:
♻️ [refactor] UploadCodyCoordinator, ProfileCoordinator, PostCoordinator 개선

dudtjs1021ej reviewed (commented):
@leeari95 
아리 꼼꼼하게 리뷰해주셔서 감사해요!!!!🥹
메모리 누수와 레이아웃 에러는 체크할 생각을 못하고 있었네요...!!
의견 반영한 것들 커밋해시 남겨놨습니다😁
혹시 빠트리거나 문제 있는 부분 있으면 말해주세요~!!

추가로 아직 해결 못한 메모리 누수가 있어요..😂
내 게시물 수정하기 탭하면 뜨는 UploadCodyVC가 계속 남아있네요..
이 부분은 좀 더 고민해보겠습니다😭

dudtjs1021ej committed:
♻️ [refactor] PostCoordinator, ProfileCoordinator 개선

- PresentType과 ProfileType을 이니셜라이저 초기화로 개선

leeari95 commented on Fitfty/Projects/Coordinator/Sources/Main/UploadCodyCoordinator.swift:
```diff
@@ -48,12 +48,13 @@ private extension UploadCodyCoordinator {
-    func dismissUploadCody(_ viewController: UIViewController) {
+    
+    func dismiss() {
+        navigationController.dismiss(animated: true)
         finishDelegate?.coordinatorDidFinish(childCoordinator: self)
```
57번 아래에 navigationController의 viewControllers를 모두 제거해주는 코드를 추가하면
메모리 누수를 해결할 수 있을 것 같아요.
네비게이션 컨트롤러 내부 childViewController가 남아있어서 메모리에서 해제되지 않는 것 같습니다!
탭바에서 새 코디 등록을 닫을 때도 동일하게 처리해주고 있어요 ㅎㅎ

```swift
// TabCoordinator.swift 205번 라인...
extension TabCoordinator: CoordinatorFinishDelegate {
    
    func coordinatorDidFinish(childCoordinator: Coordinator) {
        childDidFinish(childCoordinator, parent: self)
        switch childCoordinator.type {
        case .uploadCody:
            tabBarController.dismiss(animated: true) {
                childCoordinator.navigationController.viewControllers.removeAll()
            }
        default: break
        }
    }
}

```

leeari95 reviewed (approved):
썬~ 너무 고생하셨습니다!! 
너무너무 완벽합니다~! 최고에요. 👍
제가 마지막으로 코멘트 하나 달아두었는데 그것만 수정해주시면 될 것 같아요.

근데 제가 너무 꼼꼼하게 리뷰 드렸나... 싶네요... (썬 힘들게 한 것 같아서... ㅜㅜ) 🥲
제 의견 반영 많이 해주셔서 너무 감사합니다~🙏🏻

dudtjs1021ej committed:
🐛 [fix] UploadCodyCoordinator 메모리 누수 현상 해결

dudtjs1021ej commented on Fitfty/Projects/Coordinator/Sources/Profile/PostCoordinator.swift:
```diff
@@ -133,5 +133,13 @@ extension PostCoordinator: CoordinatorFinishDelegate {
     
     func coordinatorDidFinish(childCoordinator: Coordinator) {
         childDidFinish(childCoordinator, parent: self)
+        switch childCoordinator.type {
+        case .uploadCody:
+            navigationController.dismiss(animated: true) {
+                childCoordinator.navigationController.viewControllers.removeAll()
+            }
+        default:
+             break
+        }
```
탭바코디네이터같이 부모코디네이터에서 처리해주니 UploadCoordinator 메모리 누수 해결되네요..!👍 감사합니다~!!

dudtjs1021ej reviewed (commented):
@leeari95 
헉 아니에요 제가 더 검수하고 PR 올렸어야 했는데.. 오히려 아리가 리뷰해주시느라 고생하셨져🥹
아리 덕분에 메모리 누수 해결됐습니당~!! 감사해요ㅎㅎ
머지 할게여!!

dudtjs1021ej committed:
🔧 [chore] PostCoordinator 줄바꿈 수정
